torchvision/prototype/features/_dataset_wrapper.py

```diff
@@ -0,0 +1,388 @@
+        # we need to have access to the whole list, but `tree_flatten` would also flatten it.
+        sample_flat = tree_flatten_to_spec(sample, spec)
+        wrapped_sample_flat = [wrapper(item) for item, wrapper in zip(sample_flat, wrappers_flat)]
+        sample = tree_unflatten(wrapped_sample_flat, spec)
+
+        # We don't need to care about `transform` and `target_transform` here since `VisionDataset` joins them into a
+        # `transforms` internally:
+        # https://github.com/pytorch/vision/blob/2d92728341bbd3dc1e0f1e86c6a436049bbb3403/torchvision/datasets/vision.py#L52-L54
+        if self.transforms is not None:
+            sample = self.transforms(*sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        return len(self.vision_dataset)
+
+    @classmethod
+    def _register_wrappers_fn(cls, dataset_type: Type[datasets.VisionDataset]) -> Callable[[PyTree], PyTree]:
+        def register(wrappers_fn: PyTree) -> PyTree:
+            cls._wrappers_fns[dataset_type] = wrappers_fn
+            return wrappers_fn
+
+        return register
+
+    @classmethod
+    def from_torchvision_dataset(
+        cls,
+        dataset: datasets.VisionDataset,
+        *,
+        keep_pil_image: bool = False,
+        bounding_box_format: Optional[features.BoundingBoxFormat] = None,
+        dtypes: Optional[Dict[Type[features._Feature], Optional[torch.dtype]]] = None,
+    ) -> VisionDatasetFeatureWrapper:
+        dtypes: Dict[Type[features._Feature], Optional[torch.dtype]] = {
+            features.Image: torch.uint8,
+            features.Label: torch.int64,
+            features.Mask: torch.uint8,
+            features.BoundingBox: torch.float32,
+            features.GenericFeature: None,
+            **(dtypes or dict()),
+        }
+        wrappers_fn = cls._wrappers_fns[type(dataset)]
+        wrappers = wrappers_fn(dataset, keep_pil_image, bounding_box_format, dtypes)
+        return cls(dataset, wrappers)
+
+
+def identity_wrapper(obj: T) -> T:
+    return obj
+
+
+def generic_feature_wrapper(data: Any) -> features.GenericFeature:
+    return features.GenericFeature(data)
+
+
+def wrap_image(image: PIL.Image.Image, *, dtype: Optional[torch.dtype]) -> features.Image:
+    image = F.pil_to_tensor(image)
+    if dtype is not None:
+        image = F.convert_image_dtype(image, dtype=dtype)
+    return features.Image(image)
+
+
+def make_image_wrapper(
+    *, keep_pil_image: bool, dtypes: Dict[Type[features._Feature], Optional[torch.dtype]]
+) -> Callable[[PIL.Image.Image], Union[PIL.Image.Image, features.Image]]:
+    if keep_pil_image:
+        return identity_wrapper
+
+    return functools.partial(wrap_image, dtype=dtypes[features.Image])
+
+
+def wrap_label(label: Any, *, categories: Optional[Sequence[str]], dtype: Optional[torch.dtype]) -> features.Label:
+    return features.Label(label, categories=categories, dtype=dtype)
```

Comment:
Curious how this work for test sets where the label might be `None`?

Reply:
It doesn't so far, thanks for the heads-up!